cores/esp8266/PolledTimeout.h

```diff
@@ -45,19 +47,78 @@ struct YieldOrSkip
   static void execute() {delay(0);}
 };
 
+template <unsigned long delayMs>
+struct YieldAndDelayMs
+{
+  static void execute() {delay(delayMs);}
+};
+
 } //YieldPolicy
 
+namespace TimePolicy
+{
+
+struct TimeSourceMillis
+{
+  // time policy in milli-seconds based on millis()
+
+  using timeType = decltype(millis());
+  static timeType time() {return millis();}
+  static constexpr timeType ticksPerSecond    = 1000;
+  static constexpr timeType ticksPerSecondMax = 1000;
+};
+
+struct TimeSourceCycles
+{
+  // time policy based on ESP.getCycleCount()
+  // this particular time measurement is intended to be called very often
+  // (every loop, every yield)
+
+  using timeType = decltype(ESP.getCycleCount());
+  static timeType time() {return ESP.getCycleCount();}
+  static constexpr timeType ticksPerSecond    = F_CPU;
+  static constexpr timeType ticksPerSecondMax = 160000000; // Mhz
+};
+
+template <typename TimeSourceType, unsigned long long second_th, unsigned long long rangeCompensate = 1>
```

Comment:
I think second_th and rangeCompensate need a Big Phat Comment explaining what they are. 
BTW, if I understand this correctly, rangeCompensate should be > 1 for cases where the TimeUnit is finer than the timeSource ticksPerSecond. In the case of nanos: 1ns < 1/ticksPerSecond => rangeCompensate = 25.
If the previous is true, it seems to me that rangeCompensate could be automagically calculated. In the case of nanos, it's 2*(second_th/TimeSourceType::ticksPerSecond).